diff --git a/drivers/rtc/rtc-s3c.c b/drivers/rtc/rtc-s3c.c
--- a/drivers/rtc/rtc-s3c.c
+++ b/drivers/rtc/rtc-s3c.c
@@ -46,6 +46,11 @@ struct s3c_rtc_data {
 	void (*disable) (struct s3c_rtc *info);
 };
 
+static struct s3c_rtc_data const s3c2410_rtc_data;
+static struct s3c_rtc_data const s3c2416_rtc_data;
+static struct s3c_rtc_data const s3c2443_rtc_data;
+static struct s3c_rtc_data const s3c6410_rtc_data;
+
 static int s3c_rtc_enable_clk(struct s3c_rtc *info)
 {
 	int ret;
@@ -386,13 +391,19 @@ static int s3c_rtc_probe(struct platform_device *pdev)
 {
 	struct s3c_rtc *info = NULL;
+	const struct platform_device_id *id;
 	int ret;
 
 	info = devm_kzalloc(&pdev->dev, sizeof(*info), GFP_KERNEL);
 	if (!info)
 		return -ENOMEM;
 
 	info->dev = &pdev->dev;
 	info->data = of_device_get_match_data(&pdev->dev);
+	if (!info->data) {
+		id = platform_get_device_id(pdev);
+		if (id)
+			info->data = (const struct s3c_rtc_data *)id->driver_data;
+	}
 	if (!info->data) {
 		dev_err(&pdev->dev, "failed getting s3c_rtc_data\n");
 		return -EINVAL;
@@ -567,9 +578,28 @@ static const __maybe_unused struct of_device_id s3c_rtc_dt_match[] = {
 };
 MODULE_DEVICE_TABLE(of, s3c_rtc_dt_match);
 
+static const struct platform_device_id s3c_rtc_driver_ids[] = {
+	{
+		.name = "s3c2410-rtc",
+		.driver_data = (kernel_ulong_t)&s3c2410_rtc_data,
+	}, {
+		.name = "s3c2416-rtc",
+		.driver_data = (kernel_ulong_t)&s3c2416_rtc_data,
+	}, {
+		.name = "s3c2443-rtc",
+		.driver_data = (kernel_ulong_t)&s3c2443_rtc_data,
+	}, {
+		.name = "s3c64xx-rtc",
+		.driver_data = (kernel_ulong_t)&s3c6410_rtc_data,
+	},
+	{ }
+};
+MODULE_DEVICE_TABLE(platform, s3c_rtc_driver_ids);
+
 static struct platform_driver s3c_rtc_driver = {
 	.probe		= s3c_rtc_probe,
 	.remove		= s3c_rtc_remove,
+	.id_table	= s3c_rtc_driver_ids,
 	.driver		= {
 		.name	= "s3c-rtc",
 		.pm	= &s3c_rtc_pm_ops,
